File: packages/python/src/output_guard/strategies/fix_quotes.py
"""fix-quotes: Normalise smart quotes and convert single-quoted strings to double.

Mirror of packages/ts/src/strategies/fix-quotes.ts.
"""

from __future__ import annotations

import re

_SMART_OPEN = re.compile(r"[‘‛‹]")
_SMART_CLOSE = re.compile(r"[’›]")
_SMART_DQUOTE = re.compile(r"[“”„‟«»]")


def fix_quotes(text: str) -> str:
    result = _SMART_OPEN.sub("'", text)
    result = _SMART_CLOSE.sub("'", result)
    result = _SMART_DQUOTE.sub('"', result)

    out: list[str] = []
    i = 0
    n = len(result)
    while i < n:
        ch = result[i]
        if ch == "'":
            out.append('"')
            i += 1
            while i < n:
                c = result[i]
                if c == "\\":
                    nxt = result[i + 1] if i + 1 < n else ""
                    if nxt == "'":
                        out.append("'")
                        i += 2
                        continue
                    out.append(result[i] if i < n else "")
                    i += 1
                    if i < n:
                        out.append(result[i])
                        i += 1
                    continue
                if c == "'":
                    out.append('"')
                    i += 1
                    break
                if c == '"':
                    out.append('\\"')
                    i += 1
                    continue
                out.append(c)
                i += 1
        else:
            out.append(ch)
            i += 1
    return "".join(out)

File: packages/ts/src/strategies/fix-quotes.ts
/**
 * fix-quotes: Replace single-quoted keys/values and Unicode smart quotes with ASCII double quotes.
 * Must-isolate: runs before fix-keys and fix-inner-quotes.
 */

const SMART_OPEN = /[‘‛‹]/g;   // ' ‛ ‹
const SMART_CLOSE = /[’›]/g;         // ' ›
const SMART_DQUOTE = /[“”„‟«»]/g; // " " „ ‟ « »

export const fixQuotes = (text: string): string => {
  let result = text;
  // Replace smart quotes
  result = result
    .replace(SMART_OPEN, "'")
    .replace(SMART_CLOSE, "'")
    .replace(SMART_DQUOTE, '"');

  // Convert single-quoted strings to double-quoted.
  // Walk char by char to handle escapes correctly.
  let out = "";
  let i = 0;
  while (i < result.length) {
    const ch = result[i];
    if (ch === "'") {
      // Collect single-quoted token
      out += '"';
      i++;
      while (i < result.length) {
        const c = result[i];
        if (c === "\\") {
          const next = result[i + 1];
          if (next === "'") {
            // \' inside single-quoted string → plain ' (no backslash needed in double-quoted)
            out += "'";
            i += 2;
            continue;
          }
          // Other escape sequences: pass through
          out += result[i] ?? "";
          i++;
          out += result[i] ?? "";
          i++;
          continue;
        }
        if (c === "'") { out += '"'; i++; break; }
        if (c === '"') { out += '\\"'; i++; continue; }
        out += c;
        i++;
      }
    } else {
      out += ch;
      i++;
    }
  }
  return out;
};
